File: android/vessel-client/src/main/kotlin/dev/helm/vessel/VesselClient.kt
package dev.helm.vessel

import java.io.Closeable
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.serialization.json.*

/** RESOLVED means admission/refusal observed, NOT remote execution/cleanup complete. */
enum class JournalState { UNCERTAIN, RESOLVED }
class JournalEntry(
    val vesselId: String,
    val sessionId: String,
    val commandId: String,
    val exactRequest: String,
    val state: JournalState = JournalState.UNCERTAIN,
    val responseJson: String? = null,
) {
    override fun toString() = "JournalEntry(commandId=$commandId, state=$state)"
}

/**
 * Android adapter must commit durably before returning from insert. Atomic uniqueness is commandId.
 * If present, compare vesselId/sessionId/exactRequest byte-for-byte: throw on any mismatch, false
 * only for an identical intent. Never replace a saved payload. record updates only observation fields
 * atomically and must not regress RESOLVED to UNCERTAIN. Records contain private text: private storage,
 * no diagnostics or backups; credentials and socket IDs never belong in this journal.
 */
interface MutationJournal {
    suspend fun insert(entry: JournalEntry): Boolean
    suspend fun record(commandId: String, state: JournalState, responseJson: String?)
    suspend fun pending(vesselId: String): List<JournalEntry>
}

class VesselClient(
    private val endpoint: VesselEndpoint,
    private val journal: MutationJournal,
    private val transport: VesselTransport = OkHttpVesselTransport(endpoint),
) : Closeable {
    val state: StateFlow<ConnectionState> get() = transport.state
    suspend fun connect() = transport.connect()
    /** Whitelist prevents accidental mutation bypass of durable intent storage. */
    suspend fun read(request: JsonObject): VesselResponse {
        require(request["protocol"]?.jsonPrimitive?.int == 1)
        require(request.getValue("command").jsonObject.string("op") in READ_OPERATIONS) { "Use mutate for durable operations" }
        return transport.exchange(request.toString())
    }
    suspend fun mutate(mutation: Mutation): VesselResponse {
        val exact = mutation.request.toString()
        require(exact.toByteArray(Charsets.UTF_8).size <= MAX_FRAME_BYTES - 128) { "Request limit" }
        val entry = JournalEntry(endpoint.vesselId, mutation.sessionId, mutation.commandId, exact)
        // Process death at ANY point after this commit leaves a recoverable uncertainty, even before send.
        // Repeated button press with same ID never repeats external dispatch.
        if (!journal.insert(entry)) return recover(entry)
        val response = transport.exchange(exact)
        record(entry, response, receipt = false)
        return response
    }
    suspend fun recover(entry: JournalEntry): VesselResponse {
        require(entry.vesselId == endpoint.vesselId) { "Journal belongs to another Vessel" }
        uuid(entry.sessionId); uuid(entry.commandId)
        val response = read(Requests.receipt(entry.sessionId, entry.commandId))
        record(entry, response, receipt = true)
        return response
    }
    /** Bounded recovery page; caller chooses subsequent pages. Never resend or invoke resolve(original). */
    suspend fun recoverPending(limit: Int = 32): List<VesselResponse> {
        require(limit in 1..128)
        return journal.pending(endpoint.vesselId).take(limit).map { recover(it) }
    }
    private suspend fun record(entry: JournalEntry, response: VesselResponse, receipt: Boolean) {
        val envelope = response.result as? JsonObject
        val result = envelope?.get("result") as? JsonObject
        val identityMatches = envelope?.get("session_id")?.jsonPrimitive?.contentOrNull == entry.sessionId &&
            (result?.get("command_id")?.jsonPrimitive?.contentOrNull == entry.commandId ||
                (result?.get("request") as? JsonObject)?.get("receipt_id")?.jsonPrimitive?.contentOrNull == entry.commandId ||
                ((result?.get("record") as? JsonObject)?.get("request") as? JsonObject)?.get("receipt_id")?.jsonPrimitive?.contentOrNull == entry.commandId)
        // A successful receipt with unknown status is NOT evidence of non-delivery.
        // Known receipts can be accepted/requested/queued: admission resolved, execution still pending.
        val resolved = !response.outcomeUnknown && if (!receipt) {
            response.error != null || identityMatches
        } else {
            response.error == null && identityMatches && result?.get("status")?.jsonPrimitive?.contentOrNull in RECEIPT_STATUSES
        }
        journal.record(entry.commandId, if (resolved) JournalState.RESOLVED else JournalState.UNCERTAIN,
            obj("protocol" to 1.json(), "result" to response.result, "error" to (response.error?.json() ?: JsonNull), "outcome_unknown" to JsonPrimitive(response.outcomeUnknown)).toString())
    }
    fun observe(subscriptions: List<EventSubscription>): Flow<VesselEvent> = transport.observe(subscriptions)

    /** Canonical recent snapshot + revision-pinned page. Never append invalidation events as text.
     * A race/revision change fails; caller keeps cache stale and repeats reconciliation deliberately.
     * Large history is paged separately. Full message chunks use UTF-8 byte offsets from server.
     */
    suspend fun reconcile(sessionId: String, historyOffset: Long = 0, historyLimit: Int = 64): Reconciliation {
        val generation = state.value
        require(generation.socketId != null) { "Vessel not connected" }
        val snapshotReply = read(Requests.snapshot(sessionId)).voyageResult(sessionId)
        val snapshot = snapshotReply.getValue("result").jsonObject
        val revision = snapshot.number("revision")
        val historyReply = read(Requests.history(sessionId, historyOffset, historyLimit, revision)).voyageResult(sessionId)
        val history = historyReply.getValue("result").jsonObject
        require(snapshot.string("session_id") == sessionId && history.string("session_id") == sessionId && history.number("revision") == revision && snapshotReply.string("incarnation") == historyReply.string("incarnation")) { "Canonical identity changed; reconcile again" }
        require(state.value == generation) { "Connection changed; cache remains stale" }
        return Reconciliation(sessionId, snapshotReply.string("incarnation"), revision, snapshot.number("observation_cursor"), snapshot, history, generation)
    }
    override fun close() = transport.close()
    companion object {
        private val READ_OPERATIONS = setOf("capabilities", "catalogue", "inspect", "snapshot", "history", "message_chunk", "run_output", "events", "decisions", "receipt")
        private val RECEIPT_STATUSES = setOf("accepted", "requested", "already_terminal", "applied", "deleted", "transferred", "queued", "not_applied", "unknown_after_restart")
    }
}

/** Opaque JSON projections retain new fields instead of inventing narrowed snapshot wire shapes. */
class Reconciliation(val sessionId: String, val incarnation: String, val revision: Long, val cursor: Long,
                     val snapshot: JsonObject, val history: JsonObject, val connection: ConnectionState) {
    override fun toString() = "Reconciliation(sessionId=$sessionId, revision=$revision)"
}

fun VesselResponse.voyageResult(sessionId: String): JsonObject {
    if (!successful) throw VesselFailure("Vessel refused or could not confirm operation")
    val reply = result.jsonObject
    require(reply.string("session_id") == sessionId) { "Vessel response identity mismatch" }
    uuid(reply.string("incarnation"))
    return reply
}

File: web/resources/js/vessel-client.js
import {connectionDiagnostic as log} from './connection-diagnostics.js';
// Socket correlation is ephemeral. Durable command IDs are separate and never replayed.
export const uuid = () => crypto.randomUUID();
export const request = (op, fields = {}) => ({protocol: 1, command: {op, ...fields}});
export function voyageResult(response, session, incarnation = null) {
    if (response.protocol !== 1 || response.error != null || response.outcome_unknown !== false) throw new Error('Vessel refused or could not confirm this read.');
    const envelope = response.result;
    if (envelope?.session_id !== session || typeof envelope.incarnation !== 'string' || (incarnation && incarnation !== envelope.incarnation)) throw new Error('Voyage identity changed. Refresh before acting.');
    return envelope;
}
export function mutation(op, snapshot, incarnation, fields = {}) {
    if (!Number.isSafeInteger(snapshot.revision) || !snapshot.session_id) throw new Error('A current snapshot is required.');
    const command = {session_id: snapshot.session_id, command_id: uuid(), expected_revision: snapshot.revision, expires_at_ms: Date.now() + 60000};
    if (op !== 'submit') command.incarnation = incarnation;
    if (['steer','cancel','respond'].includes(op)) command.run_id = snapshot.run?.run_id;
    return request(op, {...command, ...fields});
}
export function resolved(response, commandId, sessionId, receipt = false) {
    if (response?.protocol !== 1 || response.outcome_unknown !== false) return false;
    if (response.error != null) return !receipt;
    const envelope = response.result, value = envelope?.result;
    const matches = envelope?.session_id === sessionId && (value?.command_id === commandId || value?.request?.receipt_id === commandId || value?.record?.request?.receipt_id === commandId);
    return Boolean(matches && (!receipt || ['accepted', 'requested', 'already_terminal', 'applied', 'deleted', 'transferred', 'queued', 'not_applied', 'unknown_after_restart'].includes(value?.status)));
}
export class IntentJournal {
    constructor(storage, vesselId) { this.storage = storage; this.key = `helm-web:intent:${vesselId}:`; }
    entries() {
        const entries = [];
        for (let i = 0; i < this.storage.length; i++) {
            const key = this.storage.key(i);
            if (!key?.startsWith(this.key)) continue;
            const value = JSON.parse(this.storage.getItem(key));
            if (!value || typeof value.command_id !== 'string' || typeof value.session_id !== 'string' || key !== this.key + value.command_id) throw new Error('Command journal is invalid. Do not resend uncertain work.');
            entries.push(value);
        }
        return entries;
    }
    prepare(command) {
        if (this.entries().length >= 128) throw new Error('Command journal full. Reconcile outstanding receipts first.');
        // Each intent has its own key: other tabs cannot overwrite a shared array.
        // Prompts, answers and conversation text never enter persistent storage.
        this.storage.setItem(this.key + command.command_id, JSON.stringify({session_id:command.session_id, command_id:command.command_id, op:command.op, created_at:Date.now()}));
    }
    settle(commandId) { this.storage.removeItem(this.key + commandId); }
}
export class VesselSocket {
    constructor(socket, onDisconnect) {
        this.socket = socket; this.pending = new Map(); this.subscriptions = new Map(); this.onDisconnect = onDisconnect;
        socket.addEventListener('message', event => {
            try {
                const frame = JSON.parse(event.data);
                if (frame.type === 'event') {
                    const listener = this.subscriptions.get(frame.subscription_id);
                    if (listener) listener(frame.event);
                    return;
                }
                if (frame.type !== 'reply') return;
                const pending = this.pending.get(frame.request_id);
                if (pending) { log('reply',{request_id:frame.request_id,op:pending.op,elapsed_ms:Date.now()-pending.started,unknown:frame.response?.outcome_unknown === true,error:frame.response?.error != null}); clearTimeout(pending.timer); this.pending.delete(frame.request_id); pending.resolve(frame.response); }
            } catch { socket.close(); }
        });
        socket.addEventListener('close', event => {
            for (const pending of this.pending.values()) { clearTimeout(pending.timer); pending.reject(new Error('Connection lost; command outcome may be unknown.')); }
            this.pending.clear(); this.subscriptions.clear();
            const reasons = ['lease expired','heartbeat timeout','upstream closed','upstream unavailable','gateway refused','gateway capacity','rate exceeded','invalid upstream frame'];
            const reason = reasons.includes(event.reason) ? event.reason : 'connection closed';
            log('socket_closed',{code:event.code,reason}); this.onDisconnect(reason);
        });
    }
    exchange(value) {
        return new Promise((resolve, reject) => {
            if (this.socket.readyState !== 1 || this.pending.size >= 16) return reject(new Error('Connection unavailable.'));
            const id = uuid();
            const timer = setTimeout(() => { log('reply_timeout',{request_id:id,op:value.command.op,pending:this.pending.size}); this.pending.delete(id); reject(new Error('Reply timed out; command outcome may be unknown.')); }, 30000);
            this.pending.set(id, {resolve, reject, timer,op:value.command.op,started:Date.now()});
            log('request',{request_id:id,op:value.command.op,pending:this.pending.size});
            try { this.socket.send(JSON.stringify({type: 'command', request_id: id, request: value})); }
            catch (error) { clearTimeout(timer); this.pending.delete(id); reject(error); }
        });
    }
    subscribe(session, incarnation, after, onEvent) {
        if (this.socket.readyState !== 1 || this.subscriptions.size >= 16 || !Number.isSafeInteger(after) || after < 0) throw new Error('Observation unavailable.');
        const id = uuid();
        this.subscriptions.set(id, onEvent);
        try {
            this.socket.send(JSON.stringify({type:'subscribe', request_id:id, request:{protocol:1, subscriptions:[{session_id:session,incarnation,after}]}}));
        } catch (error) { this.subscriptions.delete(id); throw error; }
        return () => {
            if (!this.subscriptions.delete(id) || this.socket.readyState !== 1) return;
            this.socket.send(JSON.stringify({type:'unsubscribe',subscription_id:id}));
        };
    }
    drain() {
        return new Promise(resolve => {
            const deadline = Date.now() + 30000;
            const check = () => {
                if (!this.pending.size || this.socket.readyState !== 1 || Date.now() >= deadline) {
                    this.close(); resolve();
                } else setTimeout(check, 25);
            };
            check();
        });
    }
    close() { this.socket.close(); }
}
